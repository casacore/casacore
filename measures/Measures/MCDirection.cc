//# MCDirection.cc:  MDirection conversion routines 
//# Copyright (C) 1995-1998,2000-2002,2004,2007
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes
#include <casa/Exceptions.h>
#include <casa/BasicSL/Constants.h>
#include <casa/Utilities/Assert.h>
#include <casa/Arrays/Vector.h>
#include <casa/Arrays/ArrayMath.h>
#include <measures/Measures/MCDirection.h>
#include <measures/Measures/MeasFrame.h>
#include <casa/Quanta/MVPosition.h>
#include <measures/Measures/Nutation.h>
#include <measures/Measures/MeasTable.h>

namespace casa { //# NAMESPACE CASA - BEGIN

//# Statics
uInt MCDirection::ToRef_p[N_Routes][3] = {
  {MDirection::GALACTIC,	MDirection::J2000,	0},
  {MDirection::GALACTIC,	MDirection::B1950,	2},
  {MDirection::J2000,		MDirection::GALACTIC,	0},
  {MDirection::B1950,		MDirection::GALACTIC,	2},
  {MDirection::J2000,		MDirection::B1950,	2},
  {MDirection::J2000,		MDirection::B1950_VLA,	2},
  {MDirection::B1950,		MDirection::J2000,	2},
  {MDirection::B1950_VLA,	MDirection::J2000,	2},
  {MDirection::B1950,		MDirection::B1950_VLA,	0},
  {MDirection::B1950_VLA,	MDirection::B1950,	0},
  {MDirection::J2000,		MDirection::JMEAN,	0},
  {MDirection::B1950,		MDirection::BMEAN,	2},
  {MDirection::JMEAN,		MDirection::J2000,	0},
  {MDirection::JMEAN,		MDirection::JTRUE,	0},
  {MDirection::BMEAN,		MDirection::B1950,	2},
  {MDirection::BMEAN,		MDirection::BTRUE,	2},
  {MDirection::JTRUE,		MDirection::JMEAN,	0},
  {MDirection::BTRUE,		MDirection::BMEAN,	2},
  {MDirection::J2000,		MDirection::JNAT,	0},
  {MDirection::JNAT,		MDirection::J2000,	0},
  {MDirection::B1950,		MDirection::APP,	2},
  {MDirection::APP,		MDirection::B1950,	2},
  {MDirection::APP,		MDirection::TOPO,	0},
  {MDirection::HADEC,		MDirection::AZEL,	0},
  {MDirection::HADEC,           MDirection::AZELGEO,    0},
  {MDirection::AZEL,		MDirection::HADEC,	0},
  {MDirection::AZELGEO,         MDirection::HADEC,      0},
  {MDirection::HADEC,		MDirection::TOPO,	0},
  {MDirection::AZEL,		MDirection::AZELSW,	0},
  {MDirection::AZELGEO,         MDirection::AZELSWGEO,  0},
  {MDirection::AZELSW,		MDirection::AZEL,	0},
  {MDirection::AZELSWGEO,       MDirection::AZELGEO,    0},
  {MDirection::APP,		MDirection::JNAT,	0},
  {MDirection::JNAT,		MDirection::APP,	0},
  {MDirection::J2000,		MDirection::ECLIPTIC,	0},
  {MDirection::ECLIPTIC,	MDirection::J2000,	0},
  {MDirection::JMEAN,		MDirection::MECLIPTIC,	0},
  {MDirection::MECLIPTIC,	MDirection::JMEAN,	0},
  {MDirection::JTRUE,		MDirection::TECLIPTIC,	0},
  {MDirection::TECLIPTIC,	MDirection::JTRUE,	0},
  {MDirection::GALACTIC,	MDirection::SUPERGAL,	0},
  {MDirection::SUPERGAL,	MDirection::GALACTIC,	0},
  {MDirection::ITRF,		MDirection::HADEC,	0},
  {MDirection::HADEC,		MDirection::ITRF,	0},
  {MDirection::TOPO,		MDirection::HADEC,	0},
  {MDirection::TOPO,		MDirection::APP,	0},
  {MDirection::ICRS,		MDirection::J2000,	0},
  {MDirection::J2000,		MDirection::ICRS,	0} };
uInt MCDirection::FromTo_p[MDirection::N_Types][MDirection::N_Types];
MutexedInit MCDirection::theirMutexedInit (MCDirection::doFillState);

//# Constructors
MCDirection::MCDirection() :
  MVPOS1(0), MVPOS2(0), MVPOS3(0),
  VEC61(0), VEC62(0), VEC63(0), measMath() {
    fillState();
}

//# Destructor
MCDirection::~MCDirection() {
  clearConvert();
}

//# Operators

//# Member functions

void MCDirection::getConvert(MConvertBase &mc,
			     const MRBase &inref, 
			     const MRBase &outref) {

    uInt iin  = inref.getType();
    uInt iout = outref.getType();
    if (iin != iout) {
      Bool iplan = (iin & MDirection::EXTRA);
      Bool oplan = (iout & MDirection::EXTRA);
      if (iplan) {
	if (iin != MDirection::COMET) {
	  mc.addMethod(MCDirection::R_PLANET0);
	  mc.addFrameType(MeasFrame::EPOCH);
	  mc.addMethod((iin & ~MDirection::EXTRA) + MCDirection::R_MERCURY);
	  mc.addMethod(MCDirection::R_PLANET);
	  initConvert(MCDirection::R_PLANET, mc);
	  iin = MDirection::JNAT;
	} else {
	  mc.addMethod(MCDirection::R_COMET0);
	  mc.addFrameType(MeasFrame::EPOCH);
	  mc.addFrameType(MeasFrame::COMET);
	  mc.addFrameType(MeasFrame::POSITION);
	  mc.addMethod(MCDirection::R_COMET);
	  initConvert(MCDirection::R_COMET, mc);
	  iin = MDirection::APP;
	}
      }
      if (oplan) iout = MDirection::J2000;
      Int tmp;
      while (iin != iout) {
	tmp = FromTo_p[iin][iout];
	iin = ToRef_p[tmp][1];
	mc.addMethod(tmp);
	initConvert(tmp, mc);
      }
    }
}

void MCDirection::clearConvert() {
  delete MVPOS1;     MVPOS1 = 0;
  delete MVPOS2;     MVPOS2 = 0;
  delete MVPOS3;     MVPOS3 = 0;
  delete VEC61;	     VEC61 = 0;
  delete VEC62;	     VEC62 = 0;
  delete VEC63;	     VEC63 = 0;
}

//# Conversion routines
void MCDirection::initConvert(uInt which, MConvertBase &mc) {

  if (False) initConvert(which, mc);	// Stop warning
  if (!MVPOS1)  MVPOS1 = new MVPosition();
  if (!MVPOS2)  MVPOS2 = new MVPosition();
  if (!MVPOS3)  MVPOS3 = new MVPosition();
  if (!VEC61)   VEC61 = new Vector<Double>(6);
  if (!VEC62)   VEC62 = new Vector<Double>(6);
  if (!VEC63)   VEC63 = new Vector<Double>(6);
  
  switch (which) {
    
  case J2000_JMEAN:
    measMath.createPrecession();
    mc.addFrameType(MeasFrame::EPOCH);
    break;
    
  case B1950_BMEAN:
    measMath.createPrecessionB1950();
    mc.addFrameType(MeasFrame::EPOCH);
    break;
    
  case JMEAN_J2000:
    mc.addFrameType(MeasFrame::EPOCH);
    measMath.createPrecession();
    break;
    
  case JMEAN_JTRUE:
    mc.addFrameType(MeasFrame::EPOCH);
    measMath.createNutation();
    break;
    
  case BMEAN_B1950:
    mc.addFrameType(MeasFrame::EPOCH);
    measMath.createPrecessionB1950();
    break;
    
  case BMEAN_BTRUE:
    measMath.createNutationB1950();
    mc.addFrameType(MeasFrame::EPOCH);
    break;
    
  case JTRUE_JMEAN:
    measMath.createNutation();
    mc.addFrameType(MeasFrame::EPOCH);
    break;
    
  case BTRUE_BMEAN:
    measMath.createNutationB1950();
    mc.addFrameType(MeasFrame::EPOCH);
    break;
    
  case J2000_JNAT:
    measMath.createSolarPos();
    mc.addFrameType(MeasFrame::EPOCH);
    break;
    
  case JNAT_APP:
    measMath.createAberration();
    measMath.createPrecNutat();
    mc.addFrameType(MeasFrame::EPOCH);
    break;
    
  case JNAT_J2000:
    measMath.createSolarPos();
    mc.addFrameType(MeasFrame::EPOCH);
    break;
    
  case APP_JNAT:
    measMath.createAberration();
    measMath.createPrecNutat();
    mc.addFrameType(MeasFrame::EPOCH);
    break;
    
  case B1950_APP:
    measMath.createAberrationB1950();
    measMath.createPrecNutatB1950();
    mc.addFrameType(MeasFrame::EPOCH);
    break;
    
  case APP_B1950:
    measMath.createAberrationB1950();
    measMath.createPrecNutatB1950();
    mc.addFrameType(MeasFrame::EPOCH);
    break;
    
  case HADEC_ITRF:
  case ITRF_HADEC: 
  case HADEC_AZEL:
  case HADEC_AZELGEO:
  case AZEL_HADEC:
  case AZELGEO_HADEC:
  case MECLIP_JMEAN:
  case JMEAN_MECLIP:
  case TECLIP_JTRUE:
  case JTRUE_TECLIP:
    mc.addFrameType(MeasFrame::POSITION);
    break;

  case TOPO_HADEC: 
  case HADEC_TOPO: 
  case APP_TOPO: 
  case TOPO_APP: 
    mc.addFrameType(MeasFrame::EPOCH);
    mc.addFrameType(MeasFrame::POSITION);
    break;

  default:
    break;
    
  }
}

void MCDirection::doConvert(MeasValue &in,
			    MRBase &inref,
			    MRBase &outref,
			    const MConvertBase &mc) {
  doConvert(*(MVDirection*)&in,
	    inref, outref, mc);
}

void MCDirection::doConvert(MVDirection &in,
			    MRBase &inref,
			    MRBase &outref,
			    const MConvertBase &mc) {
  Double g1, g2, g3, lengthE, tdbTime;
  // Planetary aberration factor
  Double lengthP = 0;
  MeasTable::Types planID = MeasTable::MERCURY; // to stop warning
  uInt comID = static_cast<uInt>(MDirection::APP);
  measMath.initFrame(inref, outref);
  
  for (Int i=0; i<mc.nMethod(); i++) {
    
    switch (mc.getMethod(i)) {
 
    case HADEC_ITRF:
      measMath.applyHADECtoITRF(in);
      break;
    
    case ITRF_HADEC: 
      measMath.deapplyHADECtoITRF(in);
      break;
     
    case GAL_J2000:
      measMath.applyGALtoJ2000(in);
      break;
      
    case GAL_B1950:
      measMath.applyGALtoB1950(in);
      break;
      
    case J2000_GAL:
      measMath.deapplyGALtoJ2000(in);
      break;
      
    case B1950_GAL:
      measMath.deapplyGALtoB1950(in);
      break;
      
    case J2000_B1950:
      measMath.applyJ2000toB1950(in);
      break;
      
    case J2000_B1950_VLA:
      measMath.applyJ2000toB1950_VLA(in);
      break;
    
    case B1950_J2000:
      measMath.deapplyJ2000toB1950(in);
      break;
    
    case B1950_VLA_J2000:
      measMath.deapplyJ2000toB1950_VLA(in);
      break;

    case B1950_B1950_VLA:
      measMath.deapplyJ2000toB1950(in);
      measMath.applyJ2000toB1950_VLA(in);
      break;

    case B1950_VLA_B1950:
      measMath.deapplyJ2000toB1950_VLA(in);
      measMath.applyJ2000toB1950(in);
      break;
    
    case J2000_JMEAN:
      measMath.applyPrecession(in);
      break;
    
    case B1950_BMEAN:
      measMath.applyPrecessionB1950(in);
      break;
    
    case JMEAN_J2000:
      measMath.deapplyPrecession(in);
      break;
    
    case JMEAN_JTRUE:
      measMath.applyNutation(in);
      break;
    
    case BMEAN_B1950:
      measMath.deapplyPrecessionB1950(in);
      break;
    
    case BMEAN_BTRUE:
      measMath.applyNutationB1950(in);
      break;
    
    case JTRUE_JMEAN:
      measMath.deapplyNutation(in);
      break;
    
    case BTRUE_BMEAN:
      measMath.deapplyNutationB1950(in);
      break;
    
    case J2000_JNAT:
      measMath.applySolarPos(in);
      break;
    
    case JNAT_APP:
      measMath.applyAberration(in);
      measMath.applyPrecNutat(in);
      break;
    
    case APP_JNAT:
      measMath.deapplyPrecNutat(in);
      measMath.deapplyAberration(in);
      break;
    
    case JNAT_J2000:
      measMath.deapplySolarPos(in);
      break;
    
    case B1950_APP:
      measMath.applyPrecNutatB1950(in);
      measMath.applyAberrationB1950(in);
      break;
    
    case APP_B1950:
      measMath.deapplyAberrationB1950(in);
      measMath.deapplyPrecNutatB1950(in);
      break;
    
    case TOPO_HADEC: 
      measMath.applyTOPOtoHADEC(in);
    break;
    
    case HADEC_AZEL:
      measMath.applyHADECtoAZEL(in);
    break;

    case HADEC_AZELGEO: 
      measMath.applyHADECtoAZELGEO(in);
      break;
    
    case AZEL_HADEC:
      measMath.deapplyHADECtoAZEL(in);
    break;
    
    case AZELGEO_HADEC:
      measMath.deapplyHADECtoAZELGEO(in);
      break;

    case HADEC_TOPO: 
      measMath.deapplyTOPOtoHADEC(in);
    break;
    
    case APP_TOPO: 
      measMath.applyAPPtoTOPO(in, lengthP);
      break;
    
    case R_COMET:
      if (comID == MDirection::APP) break;
    case TOPO_APP: 
      measMath.deapplyAPPtoTOPO(in, lengthP);
      break;

    case AZEL_AZELSW: 
    case AZELSW_AZEL:
    case AZELGEO_AZELSWGEO:
    case AZELSWGEO_AZELGEO: 
      measMath.applyAZELtoAZELSW(in);
      break;

    case ECLIP_J2000:
      measMath.applyECLIPtoJ2000(in);
      break;

    case J2000_ECLIP:
      measMath.deapplyECLIPtoJ2000(in);
      break;

    case MECLIP_JMEAN:
      measMath.applyMECLIPtoJMEAN(in);
      break;

    case JMEAN_MECLIP:
      measMath.deapplyMECLIPtoJMEAN(in);
      break;

    case TECLIP_JTRUE:
      measMath.applyTECLIPtoJTRUE(in);
     break;

    case JTRUE_TECLIP:
      measMath.deapplyTECLIPtoJTRUE(in);
      break;

    case GAL_SUPERGAL:
      measMath.applyGALtoSUPERGAL(in);
      break;
      
    case SUPERGAL_GAL:
      measMath.deapplyGALtoSUPERGAL(in);
      break;
 
    case ICRS_J2000:
      measMath.applyICRStoJ2000(in);
      break;
      
    case J2000_ICRS:
      measMath.deapplyICRStoJ2000(in);
      break;
   
    case R_PLANET0: {
      MDirection::Ref::frameEpoch(outref, inref).
	getTDB(tdbTime);
      *VEC62 = MeasTable::Planetary(MeasTable::EARTH, tdbTime); // Eb
      *VEC63 = MeasTable::Planetary(MeasTable::SUN, tdbTime);   // Sb
      for (Int j=0; j<3; j++)
	(*MVPOS3)(j) = (*VEC62)(j) - (*VEC63)(j);		// E
    }
    break;
    
    case R_PLANET: {
      // Correct for light time
      MVPOS3->adjust(g2);
      lengthE = 0;
      do {
	g3 = lengthE;
	*VEC61 = MeasTable::Planetary(planID, tdbTime - g3);
	*VEC63 = MeasTable::Planetary(MeasTable::SUN, tdbTime - g3); // Sb
	for (Int j=0; j<3; j++) {
	  (*MVPOS1)(j) = (*VEC61)(j) - (*VEC62)(j);		// P
	  (*MVPOS2)(j) = (*VEC61)(j) - (*VEC63)(j);		// Q
	}
	MVPOS1->adjust(lengthE);
	lengthP = Quantity(lengthE, "AU").getBaseValue();
	MVPOS2->adjust(g1);
	if (planID != MeasTable::SUN)
	  lengthE += 2*MeasTable::Planetary(MeasTable::GMS) *
	  log((g2+lengthE+g1)/(g2-lengthE+g1));
	lengthE /= MeasTable::Planetary(MeasTable::CAU);
      } while (abs(g3-lengthE) > 1e-9*lengthE);
      in = *MVPOS1; in.adjust();
      // Correct for light deflection
      // Check if near sun
      if (planID != MeasTable::SUN &&
	  (g3 = in * *MVPOS3, 
	   !nearAbs(g3, 1.0,
		    1.0-cos(MeasTable::Planetary(MeasTable::RADS)/g2)))) {
	g1 = 2*MeasTable::Planetary(MeasTable::GMS) / g2;
	g1 /= (1.0 + (*MVPOS2)*(*MVPOS3));
	in += g1 * ((in*(*MVPOS2))* *MVPOS3 - ((*MVPOS3)*in)* *MVPOS2);
      }
      in.adjust();
    }
    break;
    
    case R_MERCURY:
      planID = MeasTable::MERCURY;
      break;
    
    case R_VENUS:
      planID = MeasTable::VENUS;
      break;
    
    case R_MARS:
      planID = MeasTable::MARS;
      break;
    
    case R_JUPITER:
      planID = MeasTable::JUPITER;
      break;
    
    case R_SATURN:
      planID = MeasTable::SATURN;
      break;
    
    case R_URANUS:
      planID = MeasTable::URANUS;
      break;
    
    case R_NEPTUNE:
      planID = MeasTable::NEPTUNE;
      break;
    
    case R_PLUTO:
      planID = MeasTable::PLUTO;
      break;
    
    case R_SUN:
      planID = MeasTable::SUN;
      break;
    
    case R_MOON:
      planID = MeasTable::MOON;
      break;
    
    case R_COMET0: {
      MDirection::Ref::frameComet(inref, outref).
	getCometType(comID);
      if (!MDirection::Ref::frameComet(inref, outref).
	  getComet(*MVPOS1)) {
	throw(AipsError("No or outside range comet table specified"));
      }
      MVPOS1->adjust(lengthP);
      in = *MVPOS1;
    }
    break;

    default:
      break;
      
    }	// switch
  }	// for
}

String MCDirection::showState() {
  fillState();
  return MCBase::showState(MCDirection::FromTo_p[0],
			   MDirection::N_Types, MCDirection::N_Routes,
			   MCDirection::ToRef_p);
}

void MCDirection::doFillState (void*) {
  MDirection::checkMyTypes();
  MCBase::makeState(FromTo_p[0],  MDirection::N_Types, N_Routes, ToRef_p);
}

} //# NAMESPACE CASA - END

