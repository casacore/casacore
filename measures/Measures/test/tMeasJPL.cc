//# tMeasJPL.cc: This program test JPL DE functions
//# Copyright (C) 1997-2002,2007
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes
#include <casa/aips.h>
#include <casa/iomanip.h>
#include <casa/Arrays/Vector.h>
#include <casa/Arrays/ArrayIO.h>
#include <measures/Measures.h>
#include <measures/Measures/MeasJPL.h>
#include <casa/Quanta/MVEpoch.h>
#include <casa/Quanta/MVDirection.h>
#include <measures/Measures/MCDirection.h>
#include <measures/Measures/MeasIERS.h>
#include <measures/Measures/MeasTable.h>
#include <measures/Measures/MDirection.h>
#include <measures/Measures/MEpoch.h>
#include <casa/iostream.h>

#include <casa/namespace.h>
int main()
{
  const MVEpoch dat = 51116;
  try {
    MVDirection mvd1;

    cout << "Test measure class MeasJPL" << endl;
    cout << "---------------------------" << endl;

    cout << setprecision(9);
    Vector<Double> val(6);

    cout << "DE200: " << dat << endl;
    cout << "---------------------------" << endl;
    cout << "Mercury0:   " <<
      MeasTable::Planetary(MeasTable::MERCURY, dat.get()) << endl;
    MeasJPL::get(val, MeasJPL::DE200, MeasJPL::MERCURY, dat);
    cout << "Mercury:    " << val << endl;
    MeasJPL::get(val, MeasJPL::DE200, MeasJPL::VENUS, dat);
    cout << "Venus:      " << val << endl;
    for (uInt i=0; i<3; i++) mvd1(i) = val(i); mvd1.adjust();
    cout << "Venus:      " << mvd1 << endl;
    MeasJPL::get(val, MeasJPL::DE200, MeasJPL::EARTH, dat);
    cout << "Earth:      " << val << endl;
    MeasJPL::get(val, MeasJPL::DE200, MeasJPL::MARS, dat);
    cout << "Mars:       " << val << endl;
    MeasJPL::get(val, MeasJPL::DE200, MeasJPL::JUPITER, dat);
    cout << "Jupiter:    " << val << endl;
    MeasJPL::get(val, MeasJPL::DE200, MeasJPL::SATURN, dat);
    cout << "Saturn:     " << val << endl;
    MeasJPL::get(val, MeasJPL::DE200, MeasJPL::URANUS, dat);
    cout << "Uranus:     " << val << endl;
    MeasJPL::get(val, MeasJPL::DE200, MeasJPL::NEPTUNE, dat);
    cout << "Neptune:    " << val << endl;
    MeasJPL::get(val, MeasJPL::DE200, MeasJPL::PLUTO, dat);
    cout << "Pluto:      " << val << endl;
    MeasIERS::closeTables();
    MeasJPL::get(val, MeasJPL::DE200, MeasJPL::MOON, dat);
    cout << "Moon:       " << val << endl;
    MeasJPL::get(val, MeasJPL::DE200, MeasJPL::SUN, dat);
    cout << "SUN:        " << val << endl;
    MeasJPL::get(val, MeasJPL::DE200, MeasJPL::BARYSOLAR, dat);
    cout << "Barycentre: " << val << endl;
    MeasJPL::get(val, MeasJPL::DE200, MeasJPL::BARYEARTH, dat);
    cout << "Earth/Moon: " << val << endl;
    MeasJPL::get(val, MeasJPL::DE200, MeasJPL::NUTATION, dat);
    cout << "Nutation:   " << val << endl;
    MeasJPL::get(val, MeasJPL::DE200, MeasJPL::LIBRATION, dat);
    cout << "Libration:  " << val << endl;

    cout << "DE405: " << dat << endl;
    cout << "---------------------------" << endl;
    MeasJPL::get(val, MeasJPL::DE405, MeasJPL::MERCURY, dat);
    cout << "Mercury:    " << val << endl;
    MeasJPL::get(val, MeasJPL::DE405, MeasJPL::VENUS, dat);
    cout << "Venus:      " << val << endl;
    MeasJPL::get(val, MeasJPL::DE405, MeasJPL::EARTH, dat);
    cout << "Earth:      " << val << endl;
    MeasJPL::get(val, MeasJPL::DE405, MeasJPL::MARS, dat);
    cout << "Mars:       " << val << endl;
    MeasJPL::get(val, MeasJPL::DE405, MeasJPL::JUPITER, dat);
    cout << "Jupiter:    " << val << endl;
    MeasJPL::get(val, MeasJPL::DE405, MeasJPL::SATURN, dat);
    cout << "Saturn:     " << val << endl;
    MeasJPL::get(val, MeasJPL::DE405, MeasJPL::URANUS, dat);
    cout << "Uranus:     " << val << endl;
    MeasJPL::get(val, MeasJPL::DE405, MeasJPL::NEPTUNE, dat);
    cout << "Neptune:    " << val << endl;
    MeasJPL::get(val, MeasJPL::DE405, MeasJPL::PLUTO, dat);
    cout << "Pluto:      " << val << endl;
    MeasJPL::get(val, MeasJPL::DE405, MeasJPL::MOON, dat);
    cout << "Moon:       " << val << endl;
    MeasJPL::get(val, MeasJPL::DE405, MeasJPL::SUN, dat);
    cout << "SUN:        " << val << endl;
    MeasJPL::get(val, MeasJPL::DE405, MeasJPL::BARYSOLAR, dat);
    cout << "Barycentre: " << val << endl;
    MeasJPL::get(val, MeasJPL::DE405, MeasJPL::BARYEARTH, dat);
    cout << "Earth/Moon: " << val << endl;
    MeasJPL::get(val, MeasJPL::DE405, MeasJPL::NUTATION, dat);
    cout << "Nutation:   " << val << endl;
    MeasJPL::get(val, MeasJPL::DE405, MeasJPL::LIBRATION, dat);
    cout << "Libration:  " << val << endl;

  } catch (AipsError x) {
    cout << x.getMesg() << endl;
  } 
  
  try {
    const MEpoch mdat(dat, MEpoch::Ref(MEpoch::TDB));
    MeasFrame frame(mdat);
    MDirection::Ref venr(MDirection::VENUS, frame);
    MDirection::Ref sunr(MDirection::SUN, frame);
    MDirection::Ref moonr(MDirection::MOON, frame);
    MDirection ven(venr);
    MDirection sn(sunr);
    MDirection mon(moonr);
    MDirection::Convert vc1(ven, MDirection::Ref(MDirection::JNAT));
    MDirection::Convert vc2(ven, MDirection::Ref(MDirection::APP));
    MDirection::Convert sc1(sn, MDirection::Ref(MDirection::JNAT));
    MDirection::Convert sc2(sn, MDirection::Ref(MDirection::APP));
    MDirection::Convert mc1(mon, MDirection::Ref(MDirection::JNAT));
    MDirection::Convert mc2(mon, MDirection::Ref(MDirection::APP));

    cout << "Venus JNAT: " << vc1() << endl;
    cout << "Venus APP:  " << vc2() << endl;
    cout << "Sun   JNAT: " << sc1() << endl;
    cout << "Sun   APP:  " << sc2() << endl;
    cout << "Sun   APP:  " << sc2().getValue().getAngle("deg") << endl;
    cout << "Moon  JNAT: " << mc1() << endl;
    cout << "Moon  APP:  " << mc2() << endl;
    cout << "Moon  APP:  " << mc2().getValue().getAngle("deg") << endl;

  } catch (AipsError x) {
    cout << x.getMesg() << endl;
  } 
  
  return 0;
}
