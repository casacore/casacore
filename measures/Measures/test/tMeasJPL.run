#!/bin/sh

rm -rf tMeasJPL_tmp.out_*
$casa_checktool ./tMeasJPL > tMeasJPL_tmp.out_x
for nm in a0 a1 a2 a3 b0 b1 b2 b3 x
do
  echo ''
  echo " *** tMeasJPL_tmp.out_$nm"
  cat tMeasJPL_tmp.out_$nm
done
