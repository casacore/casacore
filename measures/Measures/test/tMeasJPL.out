Test measure class MeasJPL
---------------------------
DE200: 51116::00:00:00.0000000
---------------------------
Mercury0:   [0.112635159, -0.378671577, -0.214656378, 0.0214596316, 0.00873928166, 0.00244336875]
Mercury:    [0.112635159, -0.378671577, -0.214656378, 0.0214596316, 0.00873928166, 0.00244336875]
Venus:      [-0.590618963, -0.402576215, -0.144035251, 0.0118285844, -0.0146926142, -0.00735831801]
Venus:      [-0.810021815, -0.552125037, -0.197541397]
Earth:      [0.791134643, 0.541235184, 0.234895275, -0.0104883661, 0.0126342196, 0.0054783985]
Mars:       [-1.17789402, 1.05226562, 0.514485072, -0.00936377554, -0.00801489163, -0.00342273851]
Jupiter:    [4.94633894, -0.187963308, -0.20105115, 0.000286172646, 0.00725274959, 0.0031018692]
Saturn:     [7.99348975, 4.46636965, 1.50082979, -0.00311780692, 0.0043883827, 0.00194651371]
Uranus:     [13.2265021, -13.5166811, -6.10703428, 0.00290631103, 0.00224550355, 0.00094236082]
Neptune:    [15.681093, -23.6715664, -10.0793028, 0.00266008804, 0.001553314, 0.000569557175]
Pluto:      [-11.1730412, -27.468227, -5.20568214, 0.00298537174, -0.0012626622, -0.00129351537]
Moon:       [0.793219046, 0.539884505, 0.234327217, -0.0101744223, 0.0131311316, 0.00563522659]
SUN:        [-0.00839636156, 0.000716225798, 0.00055030648, 3.56132935e-07, -8.3589462e-06, -3.58657669e-06]
Barycentre: [0, 0, 0, 0, 0, 0]
Earth/Moon: [0.791159666, 0.54121897, 0.234888456, -0.0104845972, 0.0126401849, 0.00548028117]
Nutation:   [-5.17424774e-05, -3.70226372e-05, -2.0672392e-07, 1.19726803e-07, 1890944.74, 819838.394]
Libration:  [0, 0, 0, 0, 0, 0]
DE405: 51116::00:00:00.0000000
---------------------------
Mercury:    [0.112637744, -0.378676342, -0.21465838, 0.021459632, 0.008739282, 0.00244336657]
Venus:      [-0.590616356, -0.402580995, -0.144037238, 0.0118285851, -0.0146926137, -0.00735831838]
Earth:      [0.791137221, 0.541230466, 0.234893242, -0.0104883661, 0.012634219, 0.00547839922]
Mars:       [-1.17789144, 1.0522608, 0.514483169, -0.00936377497, -0.00801489194, -0.00342273798]
Jupiter:    [4.94634107, -0.18797011, -0.201056025, 0.000286178011, 0.00725275016, 0.00310186884]
Saturn:     [7.99349803, 4.46636103, 1.50082792, -0.00311780118, 0.00438838396, 0.00194651469]
Uranus:     [13.2264768, -13.5166788, -6.10703383, 0.0029063126, 0.00224550748, 0.000942363076]
Neptune:    [15.6810431, -23.6716735, -10.0793458, 0.0026600922, 0.00155330188, 0.000569551256]
Pluto:      [-11.1724049, -27.4679268, -5.20569128, 0.00298543173, -0.00126258865, -0.00129350676]
Moon:       [0.793221624, 0.539879787, 0.234325184, -0.0101744222, 0.0131311309, 0.00563522725]
SUN:        [-0.00839376368, 0.000711458554, 0.000548321414, 3.5653306e-07, -8.3588289e-06, -3.58654944e-06]
Barycentre: [0, 0, 0, 0, 0, 0]
Earth/Moon: [0.791162243, 0.541214252, 0.234886423, -0.0104845973, 0.0126401843, 0.0054802819]
Nutation:   [-5.17424758e-05, -3.70226373e-05, -2.06726636e-07, 1.19728605e-07, 1890944.66, 819838.502]
Libration:  [-0.0338641195, 0.431994327, 2465.69683, -0.000115944255, -6.9822504e-05, 0.230073691]
Venus JNAT: Direction: [-0.805429762, -0.550020309, -0.220817931]
Venus APP:  Direction: [-0.805689038, -0.549705008, -0.220657152]
Sun   JNAT: Direction: [-0.805053259, -0.544252201, -0.235963537]
Sun   APP:  Direction: [-0.805312853, -0.543937631, -0.235803016]
Sun   APP:  [-145.963475, -13.6389626] deg
Moon  JNAT: Direction: [0.818069535, -0.530141109, -0.222963319]
Moon  APP:  Direction: [0.817892627, -0.530380339, -0.223043374]
Moon  APP:  [-32.9622731, -12.8878482] deg
